core/src/main/scala/kafka/coordinator/TransactionLogManager.scala

```diff
@@ -17,15 +17,334 @@
 
 package kafka.coordinator
 
-import kafka.common.Topic
+import kafka.common.{KafkaException, MessageFormatter, Topic}
+import kafka.utils.CoreUtils.inLock
 import kafka.utils.{Logging, ZkUtils}
+
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.protocol.types.Type._
+import org.apache.kafka.common.protocol.types.{ArrayOf, Field, Schema, Struct}
+import org.apache.kafka.common.TopicPartition
 import org.apache.kafka.common.utils.Utils
 
-import scala.collection.mutable
 import java.util.concurrent.locks.ReentrantLock
+import java.io.PrintStream
+import java.nio.ByteBuffer
 
-import kafka.utils.CoreUtils.inLock
+import scala.collection.mutable
+
+/**
+  * Messages stored for the transaction topic has versions for both the key and value fields. Key
+  * version is used to indicate the type of the message (also to differentiate different types
+  * of messages from being compacted together if they have the same field values); and value
+  * version is used to evolve the messages within their data types:
+  *
+  * key version 0 (producer pid mapping):               [transactionalId]
```

Comment:
One of the updates from Joel's comments was to add a separate type field instead of overloading the version field like we do for the consumer coordinator messages. So the updated schema looks like this:

MappingKey => Version Type TransactionalId
StatusKey => Version Type PID

So for now, we would only implement version=0 and use type=0 for the mapping (say), and type=1 for the status. I actually feel this is a little weird because version 0 of the schema means a different schema depending on the type. Maybe we could just standardize and use the TransactionalId for both cases? Then the schema is simply this:

TxnLogKey => Version Type TransactionalId

And both types would use the same schema. We would have to add the PID to the value of the status message though. What do you think?